apps/parsley/src/utils/filterLogs/index.ts

```diff
@@ -49,26 +73,69 @@ const filterLogs = (options: FilterLogsParams): ProcessedLogLines => {
   if (matchingLines === undefined) {
-    if (sectioningEnabled && sectionData?.functions.length) {
+    if (sectioningEnabled && sectionData?.functions.length && sectionState) {
       const filteredLines: ProcessedLogLines = [];
-      let sectionIndex = 0;
+      let funcIndex = 0;
+      let commandIndex = 0;
       for (let idx = 0; idx < logLines.length; idx++) {
-        const func = sectionData.functions[sectionIndex];
-        const isSectionStart = func && idx === func.range.start;
-        if (isSectionStart && sectionState) {
-          const isOpen = sectionState[func.functionID]?.isOpen ?? false;
+        const func = sectionData.functions[funcIndex];
+        const isFuncStart = func && idx === func.range.start;
+        const isFuncOpen = sectionState[func?.functionID]?.isOpen ?? false;
+
+        const command = sectionData.commands[commandIndex];
+        const isCommandStart = command && idx === command.range.start;
+
+        if (isFuncStart) {
+          // A function start is detected.
           filteredLines.push({
             ...func,
-            isOpen,
+            isOpen: isFuncOpen,
             rowType: RowType.SectionHeader,
           });
-          sectionIndex += 1;
-          if (isOpen) {
-            filteredLines.push(idx);
+          funcIndex += 1;
+          if (isFuncOpen) {
+            if (isCommandStart) {
+              // The line marks the start of an open function and a command.
+              const commandLine = getProccessedCommandLine(
+                command,
+                sectionState,
+              );
+              filteredLines.push(commandLine);
+              if (commandLine.isOpen) {
+                filteredLines.push(idx);
+              } else {
+                // The command is closed so skip all log lines until the end of the command.
+                idx = command.range.end - 1;
+              }
+              commandIndex += 1; // Move to the next command.
+            } else {
+              throw new Error(
+                "A function does not share the same start as a command.",
+              );
+            }
           } else {
+            // The function is closed. Skip all log lines until the end of the function.
             idx = func.range.end - 1;
+            // Skip all commands until the end of the function.
+            while (
+              sectionData.commands[commandIndex] &&
+              sectionData.commands[commandIndex].functionID === func.functionID
+            ) {
+              commandIndex += 1;
+            }
+          }
+        } else if (isCommandStart) {
+          // A command start is detected.
+          const commandLine = getProccessedCommandLine(command, sectionState);
+          filteredLines.push(commandLine);
+          if (commandLine.isOpen) {
+            filteredLines.push(idx);
+          } else {
+            idx = command.range.end - 1;
           }
+          commandIndex += 1;
```

Comment:
This is the exact same logic implemented above in lines 97-110. Given that, is it even necessary to check `isFuncOpen` initially?

Reply:
Good point, I flattened the if-statements to make the algorithm easier to read. `isFuncOpen` determines if the following commands should be skipped or processed so it should be checked initially.